executor/union_scan.go

```diff
@@ -222,10 +228,79 @@ func (us *UnionScanExec) getSnapshotRow(ctx context.Context) ([]types.Datum, err
+			if !us.table.Meta().IsCommonHandle {
+				cols := us.table.Meta().Columns
+				if datumCols == nil {
+					datumCols = make(map[int64]struct {
+						*model.ColumnInfo
+						offset int
+					})
+					for i, col := range us.columns {
+						datumCols[col.ID] = struct {
+							*model.ColumnInfo
+							offset int
+						}{
+							col,
+							i,
+						}
+					}
+				}
+			INDEX:
+				for _, index := range us.table.Meta().Indices {
+					// only check primary index for pessimistic transactions
+					// check all unique indexes for optimistic transactions
+					if !index.Unique || (!index.Primary && isPessimistic) {
+						continue
+					}
+					indexDatums := make([]types.Datum, len(index.Columns))
+					for i, uniqueCol := range index.Columns {
+						if col, ok := datumCols[cols[uniqueCol.Offset].ID]; ok {
+							indexDatums[i] = row.GetDatum(col.offset, &col.FieldType)
+						} else {
+							// Skip this index check since there is not a full row.
+							// Here we have the risk of missing check operation does not use union scan.
+							// e.g. push down some aggregation functions into TiKV, so we may calculate same record twice in TiDB.
+							continue INDEX
+						}
+					}
+					checkKey, err := EncodeUniqueIndexKey(us.ctx, us.table.Meta(), index, indexDatums, us.table.Meta().ID)
+					if err != nil {
+						return nil, err
+					}
+					if _, err := us.memBufSnap.Get(context.TODO(), checkKey); err == nil {
+						flag, err := us.memBuf.GetFlags(checkKey)
+						if err != nil {
+							return nil, err
+						}
+						// I do think this fix is too tricky, but seems I have no choice, it's hard to understand how it works.
+						// Here is the basic logic for optimistic and pessimistic transactions.
+						// OPTIMISTIC:
+						// There is no unique check in some executor, so we check every unique key here.
+						// The `SelectForUpdate` statements will mark keys as locked but do not actually acquire the lock.
+						// Also the key is left in membuffer, so when we try getting it, it'll return the value.
+						// Only newly added rows need to overwrite snapshot rows, so we ignore the locked keys here.
+						// PESSIMISTIC:
+						// The unique check for pessimistic transactions is well covered.
+						// For locked keys, there should not be any transactions with duplicated entries being executed.
+						// But we still want to avoid less inconsistency behavior. Consider the following case.
+						// T1 begins pessimistic and lock the primary key and change the key value from PK1 to PK2.
+						// T2 begins pessimistic and insert a record with PK2 and got blocked.
+						// T1 commits, T2 got executed.
+						// T2 scans all records, we do not want neither duplicate entries nor T2's newly added entries here(can not be seen by this snapshot).
+						// In this case, T2 should skip the locked keys and read it's newly added entry only.
+						if flag.HasLocked() == isPessimistic {
```

Comment:
This fix is toooooo tricky, I have to add a detailed explanation here. @cfzjywxk PTAL